File: ios/HybridRNLocalizeDate.swift
import Foundation

func mapDateStyleToDateFormatterStyle(_ style: DateStyle) -> DateFormatter.Style {
    switch style {
    case .short:
        return .short
    case .medium:
        return .medium
    case .long:
        return .long
    case .full:
        return .full
    }
}

func getPreferredLocale() -> Locale {
    guard let preferredIdentifier = Locale.preferredLanguages.first else {
        return Locale.current
    }
    return Locale(identifier: preferredIdentifier)
}

class HybridRNLocalizeDate: HybridRNLocalizeDateSpec {
    public var hybridContext = margelo.nitro.HybridContext()
    public var memorySize: Int {
        return getSizeOf(self)
    }

    func localizeDateOnly(time: Double, dateStyle: DateStyle) throws -> String {
        let seconds = TimeInterval(time) / 1_000
        let date = Date(timeIntervalSince1970: seconds)

        let dateFormatter = DateFormatter()
        dateFormatter.locale = getPreferredLocale()
        dateFormatter.dateStyle = mapDateStyleToDateFormatterStyle(dateStyle)
        dateFormatter.timeStyle = .none

        return dateFormatter.string(from: date)
    }

    func localizeTimeOnly(time: Double, timeStyle: DateStyle) throws -> String {
        let seconds = TimeInterval(time) / 1_000
        let date = Date(timeIntervalSince1970: seconds)

        let dateFormatter = DateFormatter()
        dateFormatter.locale = getPreferredLocale()
        dateFormatter.dateStyle = .none
        dateFormatter.timeStyle = mapDateStyleToDateFormatterStyle(timeStyle)

        return dateFormatter.string(from: date)
    }
}

File: android/src/main/java/com/margelo/nitro/localizedate/HybridRNLocalizeDate.kt
package com.margelo.nitro.localizedate

import com.margelo.nitro.NitroModules
import java.text.DateFormat
import java.util.Date

fun mapDateStyleToDateFormat(style: DateStyle): Int {
    return when (style) {
        DateStyle.SHORT -> DateFormat.SHORT
        DateStyle.MEDIUM -> DateFormat.MEDIUM
        DateStyle.LONG -> DateFormat.LONG
        DateStyle.FULL -> DateFormat.FULL
    }
}

class HybridRNLocalizeDate : HybridRNLocalizeDateSpec() {
    override val memorySize: Long
        get() = 0L

    override fun localizeTimeOnly(time: Double, timeStyle: DateStyle): String {
        val date = Date(time.toLong())

        val timeFormatStyle = mapDateStyleToDateFormat(timeStyle)

        val dateFormatter = DateFormat.getTimeInstance(timeFormatStyle)
        return dateFormatter.format(date)
    }

    override fun localizeDateOnly(time: Double, dateStyle: DateStyle): String {
        val date = Date(time.toLong())

        val dateFormatStyle = mapDateStyleToDateFormat(dateStyle)

        val dateFormatter = DateFormat.getDateInstance(dateFormatStyle)
        return dateFormatter.format(date)
    }
}
